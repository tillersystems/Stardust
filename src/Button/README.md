# Button

### Usage

```jsx
import { Button } from '@tillersystems/stardust';
```

<!-- STORY -->

### Description

A Button comes in three possible sizes and six visual sizes.  
An Icon component can be displayed next to the text label of the button, left or right side.
For now, instance of the Icon must be passed to the Button so styling is rather free.

Button can be a simple standalone button or part of a form.

<!-- PROPS -->

### Properties

| Name           | Required |                                    Type                                     | DefaultValue |                                        Description                                         |
| -------------- | :------: | :-------------------------------------------------------------------------: | :----------: | :----------------------------------------------------------------------------------------: |
| `appearance`   |    -     | `oneOf(['default', 'primary', 'secondary', 'success', 'danger', 'google'])` |  `default`   |                          The base styling to apply to the button                           |
| `children`     |    -     |                                   `node`                                    |    `null`    |    Anything that can be rendered: numbers, strings, elements or an array (or fragment)     |
| `className`    |    -     |                                  `string`                                   |     `''`     |                           ClassName needed by styled components                            |
| `fluid`        |    -     |                                   `bool`                                    |   `false`    |                             Whether the button is fluid or not                             |
| `disabled`     |    -     |                                   `bool`                                    |   `false`    |                           Whether the button is disabled or not                            |
| `icon`         |    -     |                                   `node`                                    |    `null`    |                     Adds icon next to the text button to illustrate it                     |
| `iconPosition` |    -     |                                  `string`                                   |    `left`    |  Sets the icon position compared to the text. It will be set to the left side by default   |
| `onClick`      |    -     |                                   `func`                                    |  `() = {}`   |                               Handler of click on the button                               |
| `size`         |    -     |                   `oneOf(['small', 'default', 'large'])`                    |  `default`   | Whether the button size is small, default or large. It will be set to `default` by default |
| `type`         |    -     |                        `oneOf(['button', 'submit'])`                        |   `button`   |                        Whether it is a button or a form submission                         |

### Example

```jsx
import { Button } from '@tillersystems/stardust';

render() {
  return (
    <Button appearance="primary">Click me !</Button>
  );
}
```
