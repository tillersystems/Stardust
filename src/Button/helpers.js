import { css } from 'styled-components';
import { darken } from 'polished';

/**
 * getAppearance
 *
 * Helper function to render the right css
 * compared to the appearance given to it.
 *
 * @param {object} palette // The color's pallete of the theme.
 * @param {string} appearance // The button appearance.
 */

export const getAppearance = ({ palette }, appearance) => {
  return {
    default: css`
      color: ${palette.darkBlue};
      background: none;
      svg > path {
        fill: ${palette.darkBlue};
      }
    `,
    primary: css`
      background: ${palette.primary.default}
        linear-gradient(
          0deg,
          ${({ theme: { palette } }) => palette.whiteOpacity(0)} 0%,
          ${({ theme: { palette } }) => palette.whiteOpacity(0.1)} 100%
        );
      border: 1px solid ${palette.primary.dark};
      box-shadow: inset 0 0.2rem 0 0 ${palette.whiteOpacity(0.1)};

      /* Hovered  */
      &:hover:not([disabled]):not(:active) {
        background: ${palette.primary.default};
        box-shadow: none;
      }

      /* Activated  */
      &:active {
        background-color: ${palette.primary.dark};
        border: 1px solid ${palette.primary.darker};
        box-shadow: inset 0 0 0 0.1rem ${palette.primary.darker};
      }
    `,
    secondary: css`
      color: ${palette.spaceGrey};
      background: linear-gradient(180deg, ${palette.white} 0%, ${palette.mysticGrey} 100%);
      border: 1px solid ${palette.lightGrey};
      svg > path {
        fill: ${palette.spaceGrey};
      }
    `,
    success: css`
      background: ${palette.success.default}
        linear-gradient(0deg, ${palette.whiteOpacity(0)} 0%, ${palette.whiteOpacity(0.1)} 100%);
      border: 1px solid ${palette.success.dark};
      box-shadow: inset 0 0.2rem 0 0 ${palette.whiteOpacity(0.1)};

      /* Hovered  */
      &:hover:not([disabled]):not(:active) {
        background: ${palette.success.default};
        box-shadow: none;
      }

      /* Activated  */
      &:active {
        background-color: ${palette.success.dark};
        border: 1px solid ${palette.success.darker};
        box-shadow: inset 0 0 0 0.1rem ${palette.success.darker};
      }
    `,
    failure: css`
      background: ${palette.failure.default}
        linear-gradient(0deg, ${palette.whiteOpacity(0)} 0%, ${palette.whiteOpacity(0.1)} 100%);
      border: 1px solid ${palette.failure.dark};
      box-shadow: inset 0 0.2rem 0 0 ${palette.whiteOpacity(0.1)};

      /* Hovered  */
      &:hover:not([disabled]):not(:active) {
        background: ${palette.failure.default};
        box-shadow: none;
      }

      /* Activated  */
      &:active {
        background-color: ${palette.failure.dark};
        border: 1px solid ${palette.failure.darker};
        box-shadow: inset 0 0 0 0.1rem ${palette.failure.darker};
      }
    `,
    google: css`
      background: ${palette.googleBrandRed};
      border: 0.1rem solid ${palette.googleBrandRed};
      box-shadow: inset 0 0.2rem 0 0 ${palette.whiteOpacity(0.1)};

      /* Hovered  */
      &:hover:not([disabled]):not(:active) {
        background: ${darken(0.1, palette.googleBrandRed)};
      }
    `,
  }[appearance];
};

/**
 * getSize
 *
 * Helper function to render the right css
 * compared to the size given to him.
 *
 * @param {object} fonts // The font theme.
 * @param {string} size // The button size.
 */

export const getSize = ({ fonts }, size, icon) => {
  return {
    small: css`
      padding: ${!icon ? '0.35rem 1.4rem' : '0.35rem 1.6rem 0.35rem 0.9rem'};
    `,
    default: css`
      padding: ${!icon ? '0.75rem 2.2rem' : '0.75rem 2.4rem 0.75rem 1.4rem'};
    `,
    large: css`
      padding: ${!icon ? '0.7rem 2.4rem' : '0.7rem 2.4rem 0.7rem 1.8rem'};
      font-size: ${fonts.size.big};
      line-height: 1.5;
    `,
  }[size];
};
