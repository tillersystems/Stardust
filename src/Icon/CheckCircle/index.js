import React from 'react';
import PropTypes from 'prop-types';

/**
 * Check Circle Icon comnponent
 *
 * @return {jsx}
 */
const CheckCircle = ({ fill }) => {
  return (
    <>
      <defs>
        <path
          d="M351.5 41.7a21.3 21.3 0 0 1-17.4 39A192 192 0 1 0 448 256v-19.6a21.3 21.3 0 0 1 42.7 0v19.6A234.7 234.7 0 1 1 351.5 41.7zm133 28.7a21.3 21.3 0 0 1 0 30.1L271 314.1a21.3 21.3 0 0 1-30.2 0l-64-64a21.3 21.3 0 1 1 30.2-30.2l48.9 49L454.2 70.3a21.3 21.3 0 0 1 30.2 0z"
          id="circle"
        />
      </defs>
      <g fill="none" fillRule="evenodd">
        <mask id="check" fill="#fff" fillOpacity="0">
          <use xlinkHref="#circle" />
        </mask>
        <use fill={fill} fillRule="nonzero" xlinkHref="#circle" />
        <g mask="url(#check)" fill={fill}>
          <path d="M0 0h512v512H0z" />
        </g>
      </g>
    </>
  );
};

/**
 * PropTypes Validation
 */
const { string } = PropTypes;

CheckCircle.propTypes = {
  /**
   * Fill color
   */
  fill: string,
};

/**
 * Default props.
 */
CheckCircle.defaultProps = {
  fill: 'white',
};

/**
 * Display name
 */
CheckCircle.displayName = 'CheckCircle';

export default CheckCircle;
