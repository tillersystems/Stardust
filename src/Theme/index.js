/**
 * Theme exported to the provider
 * This is globally available in styled-components when interpolating a function like so:
 * ${(props) => props.theme}
 * Or using import { withTheme } from 'styled-components';
 */

import { breakpoints } from './breakpoints';
import { palette } from './palette';
import { dimensions } from './dimensions';
import { fonts } from './fonts';

const Theme = {
  breakpoints,
  palette,
  dimensions,
  fonts,
};

export default Theme;
