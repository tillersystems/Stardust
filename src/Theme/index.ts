import { DefaultTheme } from 'styled-components';

/**
 * Theme exported to the provider
 * This is globally available in styled-components when interpolating a function like so:
 * ${(props) => props.theme}
 * Or using import { withTheme } from 'styled-components';
 */

import { breakpoints } from './breakpoints';
import { palette } from './palette';
import { dimensions } from './dimensions';
import { fonts } from './fonts';

const Theme: DefaultTheme = {
  breakpoints,
  palette,
  dimensions,
  fonts,
};

export default Theme;
