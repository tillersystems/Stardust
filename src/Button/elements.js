import styled, { css } from 'styled-components';

import { getAppearance, getSize } from './helpers';

/* Button Container */
export const Container = styled.button`
  position: relative;

  display: inline-flex;
  justify-content: center;
  align-items: center;
  flex-wrap: nowrap;

  max-width: 100%;
  width: ${({ fluid }) => (fluid ? '100%' : 'auto')};

  padding: 0.8rem 1.6rem;

  border-radius: ${({ theme: { dimensions } }) => dimensions.radius};

  font-size: ${({ theme: { fonts } }) => fonts.size.medium};
  line-height: 1.57;

  color: ${({ theme: { palette } }) => palette.white};

  cursor: pointer;
  outline: none;

  transition: background 200ms ease, box-shadow 200ms ease;

  svg > path {
    fill: ${({ theme: { palette } }) => palette.white};
  }

  /* Button Appearance: [default, primary, secondary...] */
  ${({ theme, appearance }) => getAppearance(theme, appearance)};

  /* Button Size: [small, default, large] */
  ${({ theme, size }) => getSize(theme, size)};

  /* Button Disabled */
  ${({ disabled }) =>
    disabled &&
    css`
      cursor: not-allowed;
      opacity: 0.4;
    `}
`;

/* Icon Container
 * An icon can be set to the left or to the right.
 * We change the flex order position and add a margin
 * according to the given props [left or right].
 */
export const ContainerIcon = styled.span`
  display: flex;
  order: ${({ right }) => (right ? 1 : 0)};

  margin-right: ${({ left }) => (left ? '1rem' : 0)};
  margin-left: ${({ right }) => (right ? '1rem' : 0)};
`;
