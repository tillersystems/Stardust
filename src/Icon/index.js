import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import cuid from 'cuid';

import Theme from '../Theme';
import { Data } from './data';

const { palette } = Theme;

/**
 * Predefined sizes
 */
const SIZE = {
  small: 10,
  medium: 20,
  large: 30,
};

/**
 * Icon comnponent
 *
 * See README, run storybook or see propTypes below for documentation
 *
 * @return {jsx}
 */
const Icon = ({ className, color, name, size, title, ...restProps }) => {
  const generatedId = useMemo(() => cuid(), [title]);
  const titleElementId = `icon-title-${generatedId}`;

  /**
   * Get icon size
   *
   * @return {string}
   */
  const getIconSize = useMemo(() => {
    if (SIZE.hasOwnProperty(size)) return SIZE[size];
    return size;
  }, [size]);

  /**
   * Get color
   *
   * @return {string}
   */
  const getColor = useMemo(() => {
    if (palette.hasOwnProperty(color)) return palette[color];
    return color;
  }, [color]);

  /**
   * Get icon path
   *
   * @return {string}
   */
  const getIconPath = Data[name];

  /**
   * Icon element
   */
  const IconElement = typeof getIconPath === 'function' && getIconPath;

  const rootProps = {
    'aria-hidden': title ? undefined : true,
    'aria-labelledby': title && titleElementId,
    focusable: 'false',
    height: getIconSize,
    role: 'img',
    viewBox: '0 0 512 512',
    width: getIconSize,
    ...restProps,
  };

  return (
    <svg className={className} xmlns="http://www.w3.org/2000/svg" {...rootProps}>
      {title && <title id={titleElementId}>{title}</title>}
      <g>
        {!IconElement ? (
          <path data-testid="icon-svg-path" fill={getColor} fillRule="evenodd" d={getIconPath} />
        ) : (
          <IconElement fill={getColor} />
        )}
      </g>
    </svg>
  );
};

/**
 * PropTypes Validation
 */
const { number, oneOfType, string } = PropTypes;

Icon.propTypes = {
  /**
   * className needed by styled component
   */
  className: string,

  /**
   * Fill color
   */
  color: string,

  /**
   * Icon name
   */
  name: string.isRequired,

  /**
   * Available sizes, including custom - e.g. '20px'
   */
  size: oneOfType([number, string]),

  /**
   * Alternative text
   */
  title: string,
};

/**
 * Default props.
 */
Icon.defaultProps = {
  className: undefined,
  color: 'white',
  size: SIZE.medium,
  title: null,
};

/**
 * Display name
 */
Icon.displayName = 'Icon';

export default styled(Icon)`
  display: inline-block;
`;
