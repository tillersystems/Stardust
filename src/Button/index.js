import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { Container, ContainerIcon } from './elements';

/**
 * Button
 *
 * This component is in charge of displaying
 * a button
 *
 * @param {string} appearance // The base styling to apply to the button.
 * @param {node} children // Anything that can be rendered: numbers, strings, elements or an array (or fragment).
 * @param {string} className // ClassName needed by styled components.
 * @param {bool} fluid // Whether the button is fluid or not.
 * @param {bool} disabled // Whether the button is disabled or not.
 * @param {node} icon // Add icon node here to illustrate the text aside.
 * @param {enum} iconPosition // Set the icon position compared to the text. It will be set to the left side by default.
 * @param {function} onClick // Handler of click on the button.
 * @param {enum} size // Whether the button size is small, default or large. It will be set to default by default.
 * @param {enum} type // Whether it is a button or a form submission.
 *
 * @return {jsx}
 */

const Button = forwardRef(
  (
    {
      appearance,
      children,
      className,
      disabled,
      fluid,
      icon,
      iconPosition,
      onClick,
      size,
      type,
      ...rest
    },
    ref,
  ) => {
    return (
      <Container
        innerRef={ref}
        appearance={appearance}
        className={className}
        disabled={disabled}
        fluid={fluid}
        icon={icon}
        iconPosition={iconPosition}
        onClick={onClick}
        size={size}
        type={type}
        {...rest}
      >
        {displayIcon(iconPosition, icon)}
        {children}
      </Container>
    );
  },
);

Button.displayName = 'Button';

/**
 * displayIcon
 *
 * This function is a helper to place
 * the icon correctly inside a button.
 * He return an react element with the
 * good props: left or right.
 *
 * @param {string} side // The side where the icon should be positioned [left or right].
 * @param {jsx} icon // The icon elements.
 *
 * @return {jsx}
 */
const displayIcon = (side, icon) => {
  if (!side || !icon) return null;
  return {
    left: <ContainerIcon left>{icon}</ContainerIcon>,
    right: <ContainerIcon right>{icon}</ContainerIcon>,
  }[side];
};

/**
 * PropTypes Validation.
 */
const { bool, func, node, oneOf, string } = PropTypes;
Button.propTypes = {
  appearance: oneOf(['default', 'primary', 'secondary', 'success', 'danger', 'google']),
  children: node,
  className: string,
  disabled: bool,
  fluid: bool,
  icon: node,
  iconPosition: string,
  onClick: func,
  size: oneOf(['small', 'default', 'large']),
  type: oneOf(['button', 'submit']),
};

/**
 * Default props.
 */
Button.defaultProps = {
  appearance: 'default',
  children: null,
  className: '',
  disabled: false,
  fluid: false,
  icon: null,
  iconPosition: 'left',
  onClick: () => {},
  size: 'default',
  type: 'button',
};

const styledButton = styled(Button)``;

styledButton.displayName = 'Button';

export default styledButton;
