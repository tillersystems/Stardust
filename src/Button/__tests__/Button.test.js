import React from 'react';
import { fireEvent } from 'react-testing-library';

import Button from '..';
import Icon from '../../Icon';
import Theme from '../../Theme';

describe('<Button />', () => {
  test('should render without a problem', () => {
    const { container } = render(<Button>Text</Button>);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render with another appearance', () => {
    const { container } = render(<Button appearance="primary">Text</Button>);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render with a different size', () => {
    const { getByText } = render(<Button size="large">Text</Button>);
    const buttonNode = getByText('Text');

    expect(buttonNode).toHaveStyleRule('padding', '0.7rem 2.4rem');
    expect(buttonNode).toHaveStyleRule('font-size', Theme.fonts.size.big);
    expect(buttonNode).toHaveStyleRule('line-height', '1.5');
  });

  test('should render a fluid button', () => {
    const { getByText } = render(<Button fluid>Text</Button>);
    const buttonNode = getByText('Text');

    expect(buttonNode).toHaveStyleRule('width', '100%');
  });

  test('should render with a type submit', () => {
    const { getByText } = render(<Button type="submit">Text</Button>);
    const buttonNode = getByText('Text');

    expect(buttonNode).toHaveAttribute('type', 'submit');
  });

  test('should render a disabled button', () => {
    const { getByText } = render(<Button disabled>Text</Button>);
    const buttonNode = getByText('Text');

    expect(buttonNode).toHaveAttribute('disabled', '');
    expect(buttonNode).toHaveStyleRule('cursor', 'not-allowed');
    expect(buttonNode).toHaveStyleRule('opacity', '0.4');
  });

  test('should render a left icon', () => {
    const icon = <Icon color={Theme.palette.white} name="calendar" />;
    const iconPosition = 'left';
    const { container } = render(
      <Button icon={icon} iconPosition={iconPosition}>
        Text
      </Button>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render a right icon', () => {
    const icon = <Icon color={Theme.palette.white} name="calendar" />;
    const iconPosition = 'right';
    const { container } = render(
      <Button icon={icon} iconPosition={iconPosition}>
        Text
      </Button>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should respond to a click handler', () => {
    const spy = jest.fn();
    const { getByText } = render(<Button onClick={spy}>Text</Button>);
    const buttonNode = getByText('Text');

    fireEvent.click(buttonNode);

    expect(spy).toHaveBeenCalledTimes(1);
  });
});
