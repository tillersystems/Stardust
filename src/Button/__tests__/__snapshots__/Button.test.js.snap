// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render a left icon 1`] = `
.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  max-width: 100%;
  width: auto;
  padding: 0.75rem 1.6rem;
  border-radius: 0.4rem;
  font-size: 1.4rem;
  line-height: 1.57;
  color: hsl(0,100%,100%);
  cursor: pointer;
  outline: none;
  -webkit-transition: background 200ms ease,box-shadow 200ms ease;
  transition: background 200ms ease,box-shadow 200ms ease;
  color: hsl(213,17%,20%);
  background: none;
  padding: 0.75rem 1.4rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  margin-right: 1rem;
  margin-left: 0;
}

.c2 {
  display: inline-block;
  border-image-source: none;
  border-image-width: 0;
  padding: 0;
  font-size: 20;
  line-height: 20;
}

<button
  class="c0"
  icon="[object Object]"
  type="button"
>
  <span
    class="c1"
  >
    <div
      class="c2"
      data-testid="iconContainer"
      height="20"
      width="20"
    >
      <svg
        data-testid="iconSvg"
        height="20"
        style="display: block;"
        viewBox="0 0 512 512"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M426.667 192v-64c0-11.782-9.552-21.333-21.334-21.333h-42.666V128c0 11.782-9.552 21.333-21.334 21.333S320 139.783 320 128v-21.333H192V128c0 11.782-9.551 21.333-21.333 21.333-11.782 0-21.334-9.55-21.334-21.333v-21.333h-42.666c-11.782 0-21.334 9.55-21.334 21.333v64h341.334zm0 42.667H85.333v192c0 11.782 9.552 21.333 21.334 21.333h298.666c11.782 0 21.334-9.551 21.334-21.333v-192zM192 64h128V42.667c0-11.782 9.551-21.334 21.333-21.334 11.782 0 21.334 9.552 21.334 21.334V64h42.666c35.347 0 64 28.654 64 64v298.667c0 35.346-28.653 64-64 64H106.667c-35.347 0-64-28.654-64-64V128c0-35.346 28.653-64 64-64h42.666V42.667c0-11.782 9.552-21.334 21.334-21.334S192 30.885 192 42.667V64z"
          data-testid="iconSvgPath"
          fill="hsl(0,100%,100%)"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </span>
  Text
</button>
`;

exports[`<Button /> should render a right icon 1`] = `
.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  max-width: 100%;
  width: auto;
  padding: 0.75rem 1.6rem;
  border-radius: 0.4rem;
  font-size: 1.4rem;
  line-height: 1.57;
  color: hsl(0,100%,100%);
  cursor: pointer;
  outline: none;
  -webkit-transition: background 200ms ease,box-shadow 200ms ease;
  transition: background 200ms ease,box-shadow 200ms ease;
  color: hsl(213,17%,20%);
  background: none;
  padding: 0.75rem 1.4rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
  margin-right: 0;
  margin-left: 1rem;
}

.c2 {
  display: inline-block;
  border-image-source: none;
  border-image-width: 0;
  padding: 0;
  font-size: 20;
  line-height: 20;
}

<button
  class="c0"
  icon="[object Object]"
  type="button"
>
  <span
    class="c1"
  >
    <div
      class="c2"
      data-testid="iconContainer"
      height="20"
      width="20"
    >
      <svg
        data-testid="iconSvg"
        height="20"
        style="display: block;"
        viewBox="0 0 512 512"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M426.667 192v-64c0-11.782-9.552-21.333-21.334-21.333h-42.666V128c0 11.782-9.552 21.333-21.334 21.333S320 139.783 320 128v-21.333H192V128c0 11.782-9.551 21.333-21.333 21.333-11.782 0-21.334-9.55-21.334-21.333v-21.333h-42.666c-11.782 0-21.334 9.55-21.334 21.333v64h341.334zm0 42.667H85.333v192c0 11.782 9.552 21.333 21.334 21.333h298.666c11.782 0 21.334-9.551 21.334-21.333v-192zM192 64h128V42.667c0-11.782 9.551-21.334 21.333-21.334 11.782 0 21.334 9.552 21.334 21.334V64h42.666c35.347 0 64 28.654 64 64v298.667c0 35.346-28.653 64-64 64H106.667c-35.347 0-64-28.654-64-64V128c0-35.346 28.653-64 64-64h42.666V42.667c0-11.782 9.552-21.334 21.334-21.334S192 30.885 192 42.667V64z"
          data-testid="iconSvgPath"
          fill="hsl(0,100%,100%)"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </span>
  Text
</button>
`;

exports[`<Button /> should render with another appearance 1`] = `
.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  max-width: 100%;
  width: auto;
  padding: 0.75rem 1.6rem;
  border-radius: 0.4rem;
  font-size: 1.4rem;
  line-height: 1.57;
  color: hsl(0,100%,100%);
  cursor: pointer;
  outline: none;
  -webkit-transition: background 200ms ease,box-shadow 200ms ease;
  transition: background 200ms ease,box-shadow 200ms ease;
  background: hsl(200,74%,46%) linear-gradient( 0deg, hsla(0,100%,100%,0) 0%, hsla(0,100%,100%,0.1) 100% );
  border: 1px solid #1873a0;
  box-shadow: inset 0 0.2rem 0 0 hsla(0,100%,100%,0.1);
  padding: 0.75rem 1.6rem;
}

.c0:hover:not([disabled]):not(:active) {
  background: hsl(200,74%,46%);
  box-shadow: none;
}

.c0:active {
  background-color: #1873a0;
  border: 1px solid #14638a;
  box-shadow: inset 0 0 0 0.1rem #14638a;
}

<button
  class="c0"
  type="button"
>
  Text
</button>
`;

exports[`<Button /> should render without a problem 1`] = `
.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  max-width: 100%;
  width: auto;
  padding: 0.75rem 1.6rem;
  border-radius: 0.4rem;
  font-size: 1.4rem;
  line-height: 1.57;
  color: hsl(0,100%,100%);
  cursor: pointer;
  outline: none;
  -webkit-transition: background 200ms ease,box-shadow 200ms ease;
  transition: background 200ms ease,box-shadow 200ms ease;
  color: hsl(213,17%,20%);
  background: none;
  padding: 0.75rem 1.6rem;
}

<button
  class="c0"
  type="button"
>
  Text
</button>
`;
